File: sensor-monitor/python/sensor_monitor/priority_queue.py
import heapq


class PriorityQueue:
    """Min-heap priority queue implemented from scratch."""

    def __init__(self):
        self._heap = []

    def push(self, priority: float, item):
        """Insert an item with the given priority (lower = higher priority)."""
        self._heap.append((priority, item))
        self._sift_up(len(self._heap) - 1)

    def pop(self):
        """Remove and return (priority, item) with the lowest priority value."""
        if not self._heap:
            raise IndexError("Pop from empty priority queue")
        self._heap[0], self._heap[-1] = self._heap[-1], self._heap[0]
        result = self._heap.pop()
        if self._heap:
            self._sift_down(0)
        return result

    def peek(self):
        """View the highest-priority (lowest value) item without removing."""
        if not self._heap:
            raise IndexError("Peek on empty priority queue")
        return self._heap[0]

    def __len__(self) -> int:
        return len(self._heap)

    def _sift_up(self, index: int):
        while index > 0:
            parent = (index - 1) // 2
            if self._heap[index][0] < self._heap[parent][0]:
                self._heap[index], self._heap[parent] = self._heap[parent], self._heap[index]
                index = parent
            else:
                break

    def _sift_down(self, index: int):
        size = len(self._heap)
        while True:
            smallest = index
            left = 2 * index + 1
            right = 2 * index + 2
            if left < size and self._heap[left][0] < self._heap[smallest][0]:
                smallest = left
            if right < size and self._heap[right][0] < self._heap[smallest][0]:
                smallest = right
            if smallest != index:
                self._heap[index], self._heap[smallest] = self._heap[smallest], self._heap[index]
                index = smallest
            else:
                break


# --- stdlib equivalent ---
# import heapq
# pq = []
# heapq.heappush(pq, (priority, counter, item))
# priority, counter, item = heapq.heappop(pq)

File: sensor-monitor/rust/src/priority_queue.rs
/// A min-heap priority queue.
/// Elements must implement Ord. The smallest element is at the root.
pub struct MinHeap<T: Ord> {
    heap: Vec<T>,
}

impl<T: Ord> MinHeap<T> {
    pub fn new() -> Self {
        MinHeap { heap: Vec::new() }
    }

    pub fn push(&mut self, value: T) {
        self.heap.push(value);
        self.sift_up(self.heap.len() - 1);
    }

    pub fn pop(&mut self) -> Option<T> {
        if self.heap.is_empty() {
            return None;
        }
        let last = self.heap.len() - 1;
        self.heap.swap(0, last);
        let result = self.heap.pop();
        if !self.heap.is_empty() {
            self.sift_down(0);
        }
        result
    }

    pub fn peek(&self) -> Option<&T> {
        self.heap.first()
    }

    pub fn len(&self) -> usize {
        self.heap.len()
    }

    pub fn is_empty(&self) -> bool {
        self.heap.is_empty()
    }

    fn sift_up(&mut self, mut index: usize) {
        while index > 0 {
            let parent = (index - 1) / 2;
            if self.heap[index] < self.heap[parent] {
                self.heap.swap(index, parent);
                index = parent;
            } else {
                break;
            }
        }
    }

    fn sift_down(&mut self, mut index: usize) {
        let size = self.heap.len();
        loop {
            let mut smallest = index;
            let left = 2 * index + 1;
            let right = 2 * index + 2;
            if left < size && self.heap[left] < self.heap[smallest] {
                smallest = left;
            }
            if right < size && self.heap[right] < self.heap[smallest] {
                smallest = right;
            }
            if smallest != index {
                self.heap.swap(index, smallest);
                index = smallest;
            } else {
                break;
            }
        }
    }
}

// --- stdlib equivalent ---
// std::collections::BinaryHeap is a max-heap by default.
// For a min-heap, wrap values with std::cmp::Reverse:
//   use std::collections::BinaryHeap;
//   use std::cmp::Reverse;
//   let mut heap = BinaryHeap::new();
//   heap.push(Reverse((priority, item)));
//   let Reverse((p, item)) = heap.pop().unwrap();

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_priority_queue() {
        let mut pq = MinHeap::new();
        pq.push(3); pq.push(1); pq.push(2);
        assert_eq!(pq.peek(), Some(&1));
        assert_eq!(pq.pop(), Some(1));
        assert_eq!(pq.pop(), Some(2));
        assert_eq!(pq.pop(), Some(3));
        assert_eq!(pq.pop(), None);
    }

    #[test]
    fn test_with_tuples() {
        let mut pq: MinHeap<(i32, &str)> = MinHeap::new();
        pq.push((3, "low")); pq.push((1, "high")); pq.push((2, "medium"));
        assert_eq!(pq.pop(), Some((1, "high")));
        assert_eq!(pq.pop(), Some((2, "medium")));
        assert_eq!(pq.pop(), Some((3, "low")));
    }
}